pyrk/th_system.py

```diff
@@ -53,9 +27,209 @@ def comp_from_name(self, name):
+        return to_ret
+
     def heatgen(self, component, power, omegas):
-        to_ret = (component.power_tot)*((1-self.kappa)*power + sum(omegas))
-        return to_ret.to(units.watt)
+        """
+        calculate heat transfer by conduction(watts/m3)
+
+        :param component: name of the component
+        :type component: str
+        :param power: normalized nuclear power generated in the component
+        i.e.: power*power_tot = power (in watts)
+        :type power: float
+        """
+        return (power*component.power_tot.magnitude *
+                (1-self.kappa) + sum(omegas))/component.vol.magnitude
+
+    def conductionFVM(self, component, env, t_idx, L=0.0*units.meter,
+                      k=0.0*units.meter, A=0.0*units.meter**2):
+        """
+        compute volumetric conductive heat transfer by conduction(watts/m3)
+
+        :param component: name of the component
+        :type component: str
+        :param env: name of the environment
+        :type env: str
+        :param t_idx: time step that conduction heat is computed
+        :type t_idx: int
+        :param L: conduction distance
+        :type L: float, units meter
+        :pram k: conductivity
+        :type k: float, units w/mk
+        :return: Qond, dimemsionless quantity
+        :rtype: float
+        """
+        T_b = component.T[t_idx].magnitude
+        T_env = env.T[t_idx].magnitude
+        r_b = component.ro.magnitude
+        r_env = env.ro.magnitude
+        dr = (component.ro-component.ri).magnitude
+        k = component.k.magnitude
+        return k/r_b * (r_b * T_b - r_env * T_env)/(dr**2)
+
+    def conduction_slab(self, component, env, t_idx, L,
+                        A=0.0*units.meter**2):
+        """
+        compute volumetric heat transfer by conduction(watts/m3)
+
+        :param component: name of the component
+        :type component: str
+        :param env: name of the environment
+        :type env: str
+        :param t_idx: time step that conduction heat is computed
+        :type t_idx: int
+        :return: Qond, dimemsionless quantity
+        :rtype: float
+        """
+        T_b = component.T[t_idx].magnitude
+        T_env = env.T[t_idx].magnitude
+        num = (T_b-T_env)
+        k = component.k
+        denom = (L/(k*A)).magnitude
+        return num/denom
+
+    def advection(self, t_out, t_in, m_flow, cp):
+        ''' calculate heat transfer by advection in watts
+
+        :param t_out: outlet temperature
+        :type t_out: float
+        :param t_in: inlet temperature
+        :type t_in: float
+        :param m_flow: mass flow rate though the control volume
+        :type m_flow: float
+        :param cp: specific heat capacity of the fluid
+        :type cp: float
+        :return: dimemsionless quantity of Qadvective
+        :rtype: float
+        '''
+        if t_out > t_in:
+            return m_flow.magnitude*cp.magnitude*(t_out-t_in)
+        else:
+            return 0.0
```

Comment:
When looking through your implementation of this advection function, before getting rid of the mass transfer, I noticed that it is wrong. I see why you would do this (you think that this function will only be applied to the core, which should only be cooled by advection in normal circumstances.) However, we shouldn't introduce assumptions like that into this equation as it is not generally true for all reactors in all scenarios. If t_in>t_out, then advection should contribute an increase in temperature in the body. It shouldn't be zero.

In a loss of heat sink scenario, this assumption is not always valid (keep your mind open to reactors other than the PBFHR and consider the possibility that a control rod is dropped into the core.)